# HG changeset patch
# User Rik <rik@octave.org>
# Date 1364247169 25200
#      Mon Mar 25 14:32:49 2013 -0700
# Node ID 6964e6b92fc1e330db964a57387458b733b32510
# Parent  8097a052ec577f200dfdb768c88d681b6ec42859
hot.m: Make colormap compatible with Matlab (bug #36473).

* scripts/image/hot.m: Make colormap compatible with Matlab (bug #36473).

diff --git a/scripts/image/hot.m b/scripts/image/hot.m
--- a/scripts/image/hot.m
+++ b/scripts/image/hot.m
@@ -41,22 +41,37 @@ function map = hot (n)
     endif
   else
     print_usage ();
   endif
 
   if (n == 1)
     map = [1, 1, 1];
   elseif (n > 1)
-    x = linspace (0, 1, n)';
-    r = (x < 2/5) .* (5/2 * x) ...
-      + (x >= 2/5);
-    g = (x >= 2/5 & x < 4/5) .* (5/2 * x - 1) ...
-      + (x >= 4/5);
-    b = (x >= 4/5) .* (5 * x - 4);
+    idx = floor (3/8 * n);
+    nel = idx;
+
+    r = ones (n, 1);
+    if (nel > 0) 
+      r(1:idx, 1) = [1:nel]' / nel;
+    endif
+
+    g = zeros (n, 1);
+    g(idx+1:2*idx, 1) = r(1:idx);
+    g(2*idx+1:end, 1) = 1;
+
+    idx = floor (3/4 * n);
+    if (any (mod (n, 8) == [0, 1, 3, 6]))
+      idx++;
+    endif
+    nel = n - idx + 1;
+
+    b = zeros (n, 1);
+    b(idx:end, 1) = [1:nel]' / nel;
+
     map = [r, g, b];
   else
     map = zeros (0, 3);
   endif
 
 endfunction
 
 
